### PR TITLE
Create new functional determinism demo based on bouncing ball + gradient-based optimization (new main)

### PR DESCRIPTION
**Pull request by linfeng-z**
Create functional determinism demo and gradient-based optimization demo. Preview figure below.

optimization

### CODE REVIEW

File: experiments/backprop_advanced/bouncing_ball_backprop.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    Run B simulations through the Retriever pipeline, compute B gradients.
+
+    Returns:
+        grads [B]: dL/dθ for each θ_i
+        losses [B]: final loss values
+    """
+    pipe, ball, sink = build_batch_pipeline(theta_vals, cfg)
+
+    # Forward pass via in-process stepper
+    for _ in range(cfg.T + 5):
```
Comment (sea-bass) [inline experiments/backprop_advanced/bouncing_ball_backprop.py]: why the +5?

File: experiments/backprop_advanced/bouncing_ball_backprop.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+# =============================================================================
+# Section 2: Gradient Verification
+# =============================================================================
+
+def verify_batch_gradients(
+    theta_vals: List[float],
+    cfg: PhysicsConfig,
+    eps: float = 1e-6,
+    tol: float = 1e-4,
```
Comment (sea-bass) [inline experiments/backprop_advanced/bouncing_ball_backprop.py:222]: can you describe the difference between eps and tol in the docstring?

File: experiments/backprop_advanced/bouncing_ball_backprop.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+                        help="Skip batch vs sequential comparison")
+    parser.add_argument("--horizon", "-T", type=int, default=100,
+                        help="Simulation horizon (default: 100)")
+    args = parser.parse_args()
+
+    cfg = PhysicsConfig(g=9.81, e=0.8, dt=0.01, T=args.horizon,
+                        x_target=0.5, x_init=1.0)
+
+    B = args.batch_size
+    theta_init = [2.0 + i * 1.0 for i in range(B)]   # spread from 2.0 to 2+B
```
Comment (sea-bass) [inline experiments/backprop_advanced/bouncing_ball_backprop.py]: rather than do it this way, do you want to maybe do it with like `np.arange` so that you are always in the same range, but higher batch size means at least a denser sampling?

File: experiments/backprop_advanced/bouncing_ball_backprop.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    python bouncing_ball_backprop.py --steps 200       # more gradient descent steps
+"""
+
+from __future__ import annotations
+
+import argparse
+import os
+import sys
+import time
+from typing import Any, Dict, List, Optional, Tuple
```
Comment (sea-bass) [inline experiments/backprop_advanced/bouncing_ball_backprop.py]: Python >= 3.10 doesn't need most of these typing hints except for `Any`

* `typing.Dict[float]` can just be `dict[float]`, same with `List` and `Tuple`
* `typing.Optional[float]` can be `float | None`

File: experiments/backprop_advanced/bouncing_ball_backprop.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+        grads_batch, losses_batch = run_batch_pipeline(theta_vals, cfg)
+        batch_ms = (time.perf_counter() - t0) * 1000
+
+        # Sequential (B independent scalar pipelines)
+        t0 = time.perf_counter()
+        grads_seq = []
+        for th in theta_vals:
+            g, _ = run_batch_pipeline([th], cfg)
+            grads_seq.append(g[0].item())
+        seq_ms = (time.perf_counter() - t0) * 1000
```
Comment (sea-bass) [inline experiments/backprop_advanced/bouncing_ball_backprop.py]: might be better to do `perf_counter_ns()` and divide it by 1e-6

File: experiments/backprop_advanced/bouncing_ball_backprop.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+        # Check gradients match
+        match = all(
+            abs(grads_batch[i].item() - grads_seq[i]) < 1e-6
```
Comment (sea-bass) [inline experiments/backprop_advanced/bouncing_ball_backprop.py]: should this 1e-6 tolerance be an argument?

File: experiments/backprop_advanced/physics.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    Contact resolution uses torch.where to keep the autograd graph connected.
+    """
+    v_pred = v - cfg.g * cfg.dt
+    x_pred = x + v_pred * cfg.dt
+
+    # Differentiable contact: keep computation graph alive
+    hit = (x_pred.detach() < 0.0).float()
+    x_next = torch.where(hit > 0.5, x_pred * 0.0, x_pred)
+    v_next = torch.where(hit > 0.5, -cfg.e * v_pred, v_pred)
+    return x_next, v_next
```
Comment (sea-bass) [inline experiments/backprop_advanced/physics.py:49]: this same functionality is copy-pasted in the flow in the example. Did you mean to call this instead? Or OK that they are not the same?

File: experiments/backprop_advanced/physics.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+def finite_difference_gradient(
+    theta_val: float,
+    cfg: PhysicsConfig,
+    eps: float = 1e-6,
+) -> Tuple[float, float]:
+    """
+    Scalar finite-difference gradient dL/dθ at a single θ value.
+
+    Uses forward differences: (L(θ+ε) - L(θ)) / ε
```
Comment (sea-bass) [inline experiments/backprop_advanced/physics.py]: probably doesn't matter, since you are already checking equivalence, but is it better to do central difference where you go `-theta/2` to `+theta/2`?

File: experiments/backprop_advanced/plot_optimization.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+import numpy as np
+import matplotlib
+matplotlib.use("Agg")
+import matplotlib.pyplot as plt
+
+# Local imports
+sys.path.insert(0, os.path.dirname(__file__))
+from physics import PhysicsConfig, finite_difference_gradient
+
+project_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "../.."))
```
Comment (sea-bass) [inline experiments/backprop_advanced/plot_optimization.py]: it's much easier to do this with `pathlib`. I think it would look like:

```python
from pathlib import Path

project_root = Path(__file__).resolve().parent.parent.parent
```

and if you need it as a string, just do `project_root.as_posix()`

File: experiments/determinism_testing/bouncing_ball_benchmark.py
Comment (sea-bass) [inline experiments/determinism_testing/bouncing_ball_benchmark.py:1]: lot of copy-pasted files in these experiments, like the `PhysicsConfig` dataclass in all the files, etc. Consider consolidating?

File: experiments/determinism_testing/plot_determinism_results.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+try:
+    import matplotlib.pyplot as plt
+    import matplotlib.patches as mpatches
+except ImportError:
+    print("Error: matplotlib is required. Install with: pip install matplotlib")
+    sys.exit(1)
+
+
+def load_results(csv_path: str) -> tuple[list[float], list[float], list[float], list[float]]:
+    """Load results from CSV file."""
```
Comment (sea-bass) [inline experiments/determinism_testing/plot_determinism_results.py]: if your return is going to be a tuple like this, it _has_ to be documented!

File: experiments/determinism_testing/README.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+- `bouncing_ball_benchmark.py` - Semantic comparison (event-time vs arrival-time)
+- `bouncing_ball_pipeline.py` - Actual Pipeline.run() with Flow-based gradient computation
+- `plot_determinism_results.py` - Visualization script
+- `verify_gradient.py` - Gradient verification via finite differences
+
+## Documentation
+
+See detailed analysis and results in:
+- `/Users/zlf/ProjectsRemote/2024-Retriever/RetrieverNotes/experiments/determinism/EXPERIMENT_REPORT.md`
+- `/Users/zlf/ProjectsRemote/2024-Retriever/RetrieverNotes/experiments/determinism/SUMMARY.md`
```
Comment (sea-bass) [inline experiments/determinism_testing/README.md]: Don't use paths to your local system! Are these part of this PR?

File: experiments/model_latency/test_sequencer_logic.py
Comment (sea-bass) [inline experiments/model_latency/test_sequencer_logic.py:1]: did you want this file in this PR? seems unrelated
